src/imp/linux_raw/arch/outline/arm.rs

```diff
@@ -0,0 +1,306 @@
+        a0: ArgReg<A0>,
+        u1: usize,
+        u2: usize,
+        u3: usize,
+        u4: usize,
+        u5: usize,
+        u6: usize,
+        nr: SyscallNumber,
+    ) -> !;
+    fn rsix_reordered_syscall2(
+        a0: ArgReg<A0>,
+        a1: ArgReg<A1>,
+        u2: usize,
+        u3: usize,
+        u4: usize,
+        u5: usize,
+        u6: usize,
+        nr: SyscallNumber,
+    ) -> RetReg<R0>;
+    fn rsix_reordered_syscall2_readonly(
+        a0: ArgReg<A0>,
+        a1: ArgReg<A1>,
+        u2: usize,
+        u3: usize,
+        u4: usize,
+        u5: usize,
+        u6: usize,
+        nr: SyscallNumber,
+    ) -> RetReg<R0>;
+    fn rsix_reordered_syscall3(
+        a0: ArgReg<A0>,
+        a1: ArgReg<A1>,
+        a2: ArgReg<A2>,
+        u3: usize,
+        u4: usize,
+        u5: usize,
+        u6: usize,
+        nr: SyscallNumber,
+    ) -> RetReg<R0>;
+    fn rsix_reordered_syscall3_readonly(
+        a0: ArgReg<A0>,
+        a1: ArgReg<A1>,
+        a2: ArgReg<A2>,
+        u3: usize,
+        u4: usize,
+        u5: usize,
+        u6: usize,
+        nr: SyscallNumber,
+    ) -> RetReg<R0>;
+    fn rsix_reordered_syscall4(
+        a0: ArgReg<A0>,
+        a1: ArgReg<A1>,
+        a2: ArgReg<A2>,
+        a3: ArgReg<A3>,
+        u4: usize,
+        u5: usize,
+        u6: usize,
+        nr: SyscallNumber,
+    ) -> RetReg<R0>;
+    fn rsix_reordered_syscall4_readonly(
+        a0: ArgReg<A0>,
+        a1: ArgReg<A1>,
+        a2: ArgReg<A2>,
+        a3: ArgReg<A3>,
+        u4: usize,
+        u5: usize,
+        u6: usize,
+        nr: SyscallNumber,
+    ) -> RetReg<R0>;
+    fn rsix_reordered_syscall5(
+        a0: ArgReg<A0>,
+        a1: ArgReg<A1>,
+        a2: ArgReg<A2>,
+        a3: ArgReg<A3>,
+        a4: ArgReg<A4>,
+        u5: usize,
+        u6: usize,
+        nr: SyscallNumber,
+    ) -> RetReg<R0>;
+    fn rsix_reordered_syscall5_readonly(
+        a0: ArgReg<A0>,
+        a1: ArgReg<A1>,
+        a2: ArgReg<A2>,
+        a3: ArgReg<A3>,
+        a4: ArgReg<A4>,
+        u5: usize,
+        u6: usize,
+        nr: SyscallNumber,
+    ) -> RetReg<R0>;
+    fn rsix_reordered_syscall6(
+        a0: ArgReg<A0>,
+        a1: ArgReg<A1>,
+        a2: ArgReg<A2>,
+        a3: ArgReg<A3>,
+        a4: ArgReg<A4>,
+        a5: ArgReg<A5>,
+        u6: usize,
+        nr: SyscallNumber,
+    ) -> RetReg<R0>;
+    fn rsix_reordered_syscall6_readonly(
+        a0: ArgReg<A0>,
+        a1: ArgReg<A1>,
+        a2: ArgReg<A2>,
+        a3: ArgReg<A3>,
+        a4: ArgReg<A4>,
+        a5: ArgReg<A5>,
+        u6: usize,
+        nr: SyscallNumber,
+    ) -> RetReg<R0>;
+}
+
+// Then we define inline wrapper functions that do the reordering.
+mod reorder {
+    use super::*;
+
+    #[inline]
+    #[must_use]
+    pub(in crate::imp::linux_raw) unsafe fn syscall0_readonly(nr: SyscallNumber) -> RetReg<R0> {
+        let u = std::mem::MaybeUninit::uninit().assume_init();
```

Comment:
This is instant UB. `std::mem::uninitialized()` was depracated in favor of `MaybeUninit` because it was equivalent to `MaybeUninit::uninit().assume_init()`, which is UB for almost all return types. `MaybeUninit` being one of the only exceptions.

Reply:
Yah, this is UB is a kinda what we want. All the `u1`, `u2`, ... arguments are just here so that `nr` (syscall number) can always be at the 8th place but we don't care about what values are in those values. We could make them take `0` but that would add another instruction to set it, but because we don't care I'm not sure that worth it.

Reply:
No, you want an uninit integer, which we aren't sure yet is allowed at all. Even if it is allowed, calling `.assume_uninit()` is UB when you haven't initialized the `MaybeUninit` yet. With UB anything is allowed including compiling it correctly. A future compiler version may replace `MaybeUninit::uninit().assume_init()` with `intrinsics::unreachable()`, which makes the syscall trivially unreachable.

Reply:
On one hand, compilers are very unlikely to actually do this here; we don't do arithmetic on these values, we don't branch on them, print them, share them across threads, or observe their values in any way, and they have type `usize` where every bitpattern has an inert meaning. Compilers like LLVM sometimes generate code that does what we're doing within in their own IRs. In short, we don't do anything related to the reasons why `.assume_uninit()` has UB.

But on the other other hand, the alternative here is to just have the outline assembly move the `nr` parameter into the desired register, which is cheap and straightforward. So I filed https://github.com/bytecodealliance/rsix/issues/62 to do that.

Reply:
Oh! I now get your point. I've posted a fix https://github.com/bytecodealliance/rsix/pull/64 for `arm` and `riscv64`. Thanks for pointing this out.

Reply:
> we don't do arithmetic on these values, we don't branch on them, print them, share them across threads, or observe their values in any way

This doesn't matter: https://www.ralfj.de/blog/2020/07/15/unused-data.html

> and they have type usize where every bitpattern has an inert meaning

The question is whether integers have undef as valid bitpattern. Undef is not a random bitpattern, it is a fundamentally different value that may or may not be lowered to a random bitpattern at machine level: https://www.ralfj.de/blog/2019/07/14/uninit.html If it is decided that undef is not a valid bitpattern for integers, this code will be instant UB.

Reply:
We don't even observe the values in dead code here; there's nothing to speculate, hoist, sink, strength-reduce, or partially evaluate. But I ultimately do agree; I'm just a little sad that the RISC-V outline assembly has to be more than `ecall; ret` :-).